typescript/schedules.md

```diff
@@ -0,0 +1,110 @@
+# Schedules
+
+- [Docs](https://docs.temporal.io/workflows/#schedules)
+- gRPC API:
+  - [methods](https://github.com/temporalio/api/blob/799926c86eb13d8a9717d3561ab9b0df43796c06/temporal/api/workflowservice/v1/service.proto#L328-L370)
+  - [`request_response.proto`](https://github.com/temporalio/api/blob/799926c86eb13d8a9717d3561ab9b0df43796c06/temporal/api/workflowservice/v1/request_response.proto#L821-L957)
+  - [`schedule/v1/message.proto`](https://github.com/temporalio/api/blob/master/temporal/api/schedule/v1/message.proto)
+  - [`ScheduleOverlapPolicy` enum](https://github.com/temporalio/api/blob/master/temporal/api/enums/v1/schedule.proto)
+
+## TS API
+
+```ts
+import { ScheduleClient, ScheduleOverlapPolicy } from '@temporalio/client'
+import { addWeeks } from 'date-fns'
+import { myWorkflow } from './workflows'
+
+const client = new ScheduleClient()
+
+const schedule = await client.create({
+  id: 'biz-id',
+  spec: {
+    // every hour at minute 5
+    interval: {
+      every: '1h',
+      at: '5m',
+    },
+    // every 20 days since epoch at day 2
+    // interval: {
+    //   every: '20d',
+    //   at: '2d'
+    // }
+    exclude: {
+      // skip 11:05 pm
+      hour: 23,
+      minute: 5,
+    },
+    endAt: addWeeks(new Date(), 4), 
+    jitter: '30s',
+    timezone: 'US/Eastern',
+  },
+  action: {
+    startWorkflow: {
+      workflowId: 'biz-id',
+      type: myWorkflow,
+      args: ['sorry this is the only thing reused, chad 😄'],
+    },
+  },
+  policies: {
+    overlap: ScheduleOverlapPolicy.BUFFER_ONE,
+    catchupWindow: '2m',
+    pauseOnFailure: true,
+  },
+  state: {
+    note: 'started schedule',
+    paused: true,
+    limitedActions: 10,
+  },
+  patch: {
+    triggerImmediately: true,
+    backfill: [
+      {
+        start: new Date(),
+        end: new Date(),
+        overlap: ScheduleOverlapPolicy.ALLOW_ALL,
+      },
+    ],
+    pause: true, // redundant with state.paused above (can use either)
+  },
+  memo,
+  searchAttributes,
+})
+
+const scheduleDescription = await schedule.describe()
+
+const matchingStartTimes = await schedule.listMatchingTimes({ start: new Date(), end: new Date() })
+
+await schedule.update({
+  spec,
+  action,
+  policies,
+  state,
+  conflictToken: scheduleDescription.conflictToken,
```

Comment:
@dnr I'm wondering why we don't support partial updates instead of needing the conflict token.
Was this not possible to do on the server side?

Reply:
@lorensr let's make a `partialUpdate` method in the high level API that uses the conflict token internally.
This might require retrying on conflict up to some configured threshold.

Reply:
It also requires reading the configuration if a schedule handle is obtained with `getHandle`.
Both `partialUpdate` and `getHandle` names TBD.

Reply:
You mean user does this:
```ts
const schedule = client.getHandle('schedule-biz-id')
await schedule.partialUpdate({
  spec,
})
```
and inside `partialUpdate` we do something like this?
```ts
function partialUpdate(partial) {
  const description = await this.describe()
  const update = {
    ...description.schedule,
    conflictToken: description.conflictToken,
    ...partial
  }
  await this.update(update)
  // if update fails due to old conflictToken, retry describe & update in loop for some configurable period/maxAttempts
}
```

Reply:
@lorensr it would be something like that but with retries.

Reply:
Definitely a valid option @cretz.
Either this or `partialUpdate` work for me.
Not sure if I'd want to keep `update` in the high level API at all.

Reply:
```ts
// usage:
const scheduleHandle = client.getHandle('schedule-biz-id')
await scheduleHandle.update(schedule => {
  schedule.spec.intervals[0].every = '1d';
  // unset with:
  // delete schedule.spec.intervals;
  
  // return false to stop retrying early 
}, { retry: retryPolicy })

// impl:
async function update(mutateFn, { retry }) {
  const description = await this.describe()
  let result = mutateFn(description.schedule)
  if (result instanceof Promise) {
    let result = await result
  }
  if (result === false) {
    // stop
  }
  await this.service.updateSchedule({
    schedule: description.schedule,
    conflictToken: description.conflictToken,
  })
  // if updateSchedule fails due to old conflictToken, goto top of fn, according to retry policy
}
```